### PR TITLE
Implement Fibonacci heap

### PR DESCRIPTION
**Pull request by nikox94**

### CODE REVIEW

**Comment by aviary2-wf** [comment]
## Raven
Number of Findings: 0

File: fibheap/fibheap.go [hunk 1]
```diff
@@ -0,0 +1,392 @@
...
+)
+
+/******************************************
+ ************** INTERFACE *****************
+ ******************************************/
+
+// The FloatingFibonacciHeap is an implementation of a fibonacci heap
+// with only floating-point priorities and no user data attached.
+type FloatingFibonacciHeap interface {
+	// Adds and element to the heap
```
Comment (dustinhiatt-wf) [inline fibheap/fibheap.go]: We typically follow golang's documentation recommendation found here: https://blog.golang.org/godoc-documenting-go-code.  Specifically, the docstring should start with the name of the function.

File: fibheap/fibheap.go [hunk 2]
```diff
@@ -0,0 +1,392 @@
...
+	"math"
+)
+
+/******************************************
+ ************** INTERFACE *****************
+ ******************************************/
+
+// The FloatingFibonacciHeap is an implementation of a fibonacci heap
+// with only floating-point priorities and no user data attached.
+type FloatingFibonacciHeap interface {
```
Comment (benechols-wf) [inline fibheap/fibheap.go]: Why an interface instead of just directly exposing the a fib heap struct with the same public methods?

Comment (nikox94) [inline fibheap/fibheap.go]: Hmm... I am not sure. Perhaps I don't really *need* that layer of abstraction, but.. I don't think it hurts either. I am also thinking of making the Priority variable of Entry exported... I had forgotten that people might want to use it.

Comment (nikox94) [inline fibheap/fibheap.go]: Btw. what do Aviary and Rosie do?

Comment (benechols-wf) [inline fibheap/fibheap.go]: as far as the interfaces go, check out this blog talking about design in go: https://medium.com/@cep21/what-accept-interfaces-return-structs-means-in-go-2fe879e25ee8#.geu8ysx4g

Aviary and Rosie are part of our continuous integration systems. 
Rosie runs the checks to validate a PR is ready and aviary is part of a watcher that makes system to watch for changes to sensitive files (and thus needing special reviews)

File: fibheap/fibheap.go (hunk 1 above)
Comment (nikox94) [inline fibheap/fibheap.go]: Done :)

File: fibheap/fibheap.go (hunk 2 above)
Comment (nikox94) [inline fibheap/fibheap.go]: :+1: about the blog article. Am going through it now and will change the code soon.

File: fibheap/Documentation/attribution.txt [hunk 3]
```diff
@@ -0,0 +1,6 @@
+All rights belong to Keith Schwarz (htiek@cs.stanford.edu) and Stanford University.
```
Comment (nikox94) [inline fibheap/Documentation/attribution.txt]: @joshhayes-sheen-wf: I've written a small attribution file. I'm still not sure it's acceptable to have them here like this though. I will ask the professor.

**Comment by nikox94** [comment]
I have not yet had time to remove the interface and return the struct. I will do this ASAP and will ping you guys.

Thanks,
Nick

**Comment by nikox94** [comment]
I seem to have addressed the Interface -> struct change. Could we merge soon?

File: fibheap/Documentation/FibonacciHeap.java [hunk 4]
```diff
@@ -0,0 +1,580 @@
+ /***********************************************************************
```
Comment (dustinhiatt-wf) [inline fibheap/Documentation/FibonacciHeap.java]: Is this here purely as a reference example?  If so, you might just put a link to this source code in the documentation and not include the file.

File: fibheap/fibheap.go [hunk 5]
```diff
@@ -0,0 +1,384 @@
...
+		 * reparent operation that merged two different trees of equal
+		 * priority, we need to make sure that the min pointer points to
+		 * the root-level one.
+		 */
+		if curr.Priority <= heap.min.Priority {
+			heap.min = curr
+		}
+	}
+
+	// All done. Return minimum element and no error
```
Comment (dustinhiatt-wf) [inline fibheap/fibheap.go]: This comment doesn't really add any value.

File: fibheap/fibheap.go [hunk 6]
```diff
@@ -0,0 +1,384 @@
...
+	return min, nil
+}
+
+// DecreaseKey decreases the key of the
```
Comment (dustinhiatt-wf) [inline fibheap/fibheap.go]: Why the short lines here?

File: fibheap/fibheap.go [hunk 7]
```diff
@@ -0,0 +1,384 @@
...
+// given element, sets it to the new
+// given priority and returns the node
+// if successfully set
+func (heap *FloatingFibonacciHeap) DecreaseKey(node *Entry, newPriority float64) (*Entry, error) {
+
+	if heap.IsEmpty() {
+		return nil, fmt.Errorf("Cannot decrease key in an empty heap")
```
Comment (dustinhiatt-wf) [inline fibheap/fibheap.go]: You might consider returning typed errors here.  It gives the caller an easy way to differentiate between errors.

File: fibheap/fibheap.go [hunk 8]
```diff
@@ -0,0 +1,384 @@
...
+
+	decreaseKeyUnchecked(heap, node, newPriority)
+	return node, nil
+}
+
+// Delete deletes the given element in the heap
+func (heap *FloatingFibonacciHeap) Delete(node *Entry) error {
+
+	if heap.IsEmpty() {
+		return fmt.Errorf("Cannot delete element from an empty heap")
```
Comment (dustinhiatt-wf) [inline fibheap/fibheap.go]: This is the same error message that's returned from DecreaseKey.  Another reason to extract this into a single error type.

File: fibheap/fibheap_test.go [hunk 9]
```diff
@@ -0,0 +1,427 @@
...
+	heap := NewFloatFibHeap()
+
+	heap.Enqueue(0)
+	heap.DequeueMin()
+
+	// Heap should be empty at this point
+
+	min, err := heap.Min()
+
+	assert.EqualError(t, err, "Trying to get minimum element of empty heap")
```
Comment (dustinhiatt-wf) [inline fibheap/fibheap_test.go]: This is an example where a typed error would be handy.  Checking the string means that this code can never change the error message and is quite clumsy.

File: fibheap/fibheap_test.go [hunk 10]
```diff
@@ -0,0 +1,427 @@
...
+		}
+	}
+
+	assert.NotNil(t, e1)
+	assert.NotNil(t, e2)
+	assert.NotNil(t, e3)
+
+	var err error
+
+	err = heap.Delete(e1)
```
Comment (dustinhiatt-wf) [inline fibheap/fibheap_test.go]: Should these errors be checked?

File: fibheap/fibheap_test.go [hunk 11]
```diff
@@ -0,0 +1,427 @@
...
+func BenchmarkFibHeap_DecreaseKey(b *testing.B) {
+	heap := NewFloatFibHeap()
+
+	sliceFlt := make([]float64, 0, b.N)
+	sliceE := make([]*Entry, 0, b.N)
+	for i := 0; i < b.N; i++ {
+		sliceFlt = append(sliceFlt, 2*1E10*(float64(i)-0.5))
+		sliceE = append(sliceE, heap.Enqueue(sliceFlt[i]))
+	}
+
```
Comment (dustinhiatt-wf) [inline fibheap/fibheap_test.go]: If you add b.ResetTimer you can narrow down the timing to the one operation you care about.

File: fibheap/fibheap_test.go [hunk 12]
```diff
@@ -0,0 +1,427 @@
...
+	// Delete runs in log(N) time
+	// so safe to reset timer here
+	b.ResetTimer()
+	for i := 0; i < b.N; i++ {
+		err := heap.Delete(sliceE[i])
+		assert.NoError(b, err)
```
Comment (dustinhiatt-wf) [inline fibheap/fibheap_test.go]: One thing you might consider using here is require.NoError(b, err).  Require is also in the testify lib.

File: fibheap/fibheap.go [hunk 13]
```diff
@@ -0,0 +1,384 @@
...
+	if heap.min == nil {
+		// If there are no entries left, we're done.
+		return min, nil
+	}
+
+	treeSlice := make([]*Entry, 0, heap.size)
+	toVisit := make([]*Entry, 0, heap.size)
+
+	for curr := heap.min; len(toVisit) == 0 || toVisit[0] != curr; curr = curr.next {
+		toVisit = append(toVisit, curr)
```
Comment (dustinhiatt-wf) [inline fibheap/fibheap.go]: Would it be possible to eliminate this loop and thus the toVisit allocation?  Seems like there may be some combination of checks in the outer loop immediately proceeding this one that could avoid it (although I could be wrong).

**Comment by nikox94** [comment]
Great comment, @dustinhiatt-wf. I will sit down to address them ASAP.

Regards,
Nick

File: fibheap/fibheap.go (hunk 6 above)
Comment (nikox94) [inline fibheap/fibheap.go]: fixed

File: fibheap/fibheap_test.go (hunk 9 above)
Comment (nikox94) [inline fibheap/fibheap_test.go]: :100:

File: fibheap/fibheap_test.go (hunk 10 above)
Comment (nikox94) [inline fibheap/fibheap_test.go]: Hmm.. yes... you're right.

File: fibheap/fibheap_test.go (hunk 12 above)
Comment (nikox94) [inline fibheap/fibheap_test.go]: :100:

File: fibheap/fibheap_test.go (hunk 11 above)
Comment (nikox94) [inline fibheap/fibheap_test.go]: I will need to rethink the benchmarks as a whole. they are designed a bit iffy at the moment.

File: fibheap/fibheap_test.go (hunk 9 above)
Comment (nikox94) [inline fibheap/fibheap_test.go]: It's not that clumsy and people do do that, but.. I suppose what you propose is more sophisticated and elegant.

File: fibheap/fibheap.go [hunk 14]
```diff
@@ -0,0 +1,399 @@
...
+	degree                    int
+	marked                    bool
+	next, prev, child, parent *Entry
+	// Priority is the numerical priority of the node
+	Priority float64
+}
+
+// EmptyHeapError fires when the heap is empty and an operation could
+// not be completed for that reason. Its string holds additional data.
+type EmptyHeapError string
```
Comment (dzyp) [inline fibheap/fibheap.go]: This is just a nit, but you could also just do `var ErrEmptyHeap = errors.New("message")`

File: fibheap/fibheap.go [hunk 15]
```diff
@@ -0,0 +1,399 @@
...
+
+	return min, nil
+}
+
+// DecreaseKey decreases the key of the given element, sets it to the new
+// given priority and returns the node if successfully set
+func (heap *FloatingFibonacciHeap) DecreaseKey(node *Entry, newPriority float64) (*Entry, error) {
+
+	if heap.IsEmpty() {
+		return nil, EmptyHeapError("Cannot decrease key in an empty heap")
```
Comment (dzyp) [inline fibheap/fibheap.go]: If you made this a var instead, you can eliminate copying these strings.

**Comment by dzyp** [comment]
@nikox94 Only one final comment about the errors but otherwise looks good!

**Comment by nikox94** [comment]
@dustinhiatt-wf , for both of your comments - I have basically defined custom types, so that a sophisticated client can do Type-checking of the error. The message is only for additional information and is different across methods in order to provide more context of the errors. I feel it's better to have things like that? I could externalise the constants but again, they are different across different methods in order to provide more context.

**Comment by beaulyddon-wf** [comment]
+1

**Pull request merged**